File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        return minHeapMethod(nums, k);
    }

    private int[] minHeapMethod(int[] nums, int k) {
        // count the elements
        Map<Integer, Integer> count = new HashMap<>();
        for (int n : nums) {
            count.put(n, count.getOrDefault(n, 0) + 1);
        }

        // use min-heap to narrow top k
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        for (Map.Entry<Integer, Integer> entry : count.entrySet()) {
            heap.offer(new int[]{
                entry.getKey(), entry.getValue()
            });
            if (heap.size() > k) {
                heap.poll();
            }
        }

        // translate back into an array
        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = heap.poll()[0];
        }
        return result;
    }
}

File: Data Structures & Algorithms/majority-element/submission-1.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        count = dict()
        max_count = max_n = 0

        for n in nums:
            count[n] = count.get(n, 0) + 1
            if count[n] > max_count:
                max_count = count[n]
                max_n = n
        
        return max_n
            
        